task/move.go

```diff
@@ -109,21 +114,33 @@ func MaybeMove(fetch MaybeFetcher, put MaybePutter) error {
 // early then MoveCtx returns ctx.Err() just as MaybeMove returns any errors
 // that it encounters.
 func MoveCtx(ctx context.Context, fetch Fetcher, put Putter) error {
-	// [TODO]
+	ch := make(chan Thing)
+	ctx, cancelFn := context.WithCancel(ctx)
+	defer cancelFn()
 
 	go func() {
-		// [TODO]
-		t, ok := fetch()
-		_, _ = t, ok // [Remove later]
-		// [TODO]
-	}()
+		defer close(ch)
+		for {
+			t, ok := fetch()
+			if !ok {
+				fmt.Println("===> I am done!")
+				break
+			}
 
-	var t Thing // [Remove later]
-	// [TODO]
-	put(t)
-	// [TODO]
+			select {
+			case <-ctx.Done():
+				fmt.Println(ctx.Err())
```

Comment:
There are a few ways you can get this error returned by MoveCtx(). I would recommend declaring a variable outside the goroutine and then returning it at the end since it's the simplest way.

```
var fetchErr error
go func() {
  …
    case <-ctx.Done():
      fetchErr = ctx.Err()
      return
  …
}

for thing := range ch {
  put(thing)
}

return fetchErr
```

The biggest worry with something like this is a [race condition](https://en.wikipedia.org/wiki/Race_condition#Software). What happens if the goroutine tries to set the value of `fetchErr` at the "same time" as we return it? I put "same time" in quotes because it can never technically be at the identical time and that is the exact problem with race conditions because we don't know the order of reads and writes.

It is ok here though because we can look through the code for guarantees. The `for` loop will continue until `ch` is closed because that is the definition of `range` over a channel. We also know that the channel won't be closed until the `defer close(ch)` is run, and we know that _that_ won't happen until the goroutine returns. Therefore we have proof that it is safe here.

Later we can look at other ways of returning errors, such as other channels or specific packages.